### ISSUE TITLE
salty

### ISSUE THREAD

**Issue by carlos8f** [dchest/nacl-stream-js#4]
Hi dchest,

I recently found your TweetNaCl port and it inspired me to resurrect my project 'salty' and port it from node-sodium. TweetNacl is working great and I love how compact it is!

If you're interested I'd love feedback on [salty](https://www.npmjs.com/package/salty). In contrast to your project, I'm using `nacl.lowlevel.crypto_stream_xor(output, 0, a(data), 0, data.length, a(nonce), a(k));` to XOR each stream packet, instead of using boxes. see: https://nacl.cr.yp.to/stream.html In practice I think it's safe, as long as the nonce is incremented after each packet, and most importantly it's extremely fast. Last night I encrypted/decrypted and MAC'd an 8GB mp4 file on a 7 year old mac mini in only a few minutes! I was pretty floored because I'm used to crypto programs taking ages. It seems like the age of Javascript is truly upon us :)

Cheers,
Carlos

**Comment by anri-asaturov** [comment]
@carlos8f did you try nacl-stream-js first? Is it slower on the same task?

**Comment by dchest** [comment]
@carlos8f I haven't read you code, but if I understood you correctly, you replaced authenticated encryption (boxes, XSalsa20+Poly1305) with plain stream cipher (XSalsa20). It is **not safe**. It's [malleable](https://en.wikipedia.org/wiki/Malleability_(cryptography)). Sure, it's faster, because it's missing one critical piece — Poly1305.

NaCl is designed with boxes specifically to prevent people from using plain encryption without authentication, and this is a good example why `lowlevel` interface is not documented. Perhaps, I should disallow access to it.

**Comment by carlos8f** [comment]
@dchest before encryption I sha256 the entire stream (concat'ed with the nonce), and that is used for authentication of the entire plaintext or else an error is thrown and the invalid plaintext is discarded. This way I protect against malleability but it requires pre-processing the full stream (suitable for file-based operations but not network). Further, while it uses mac-then-encrypt, I don't believe it's subject to the [doom principle](https://moxie.org/blog/the-cryptographic-doom-principle/) because there is only one error state (output didn't match the sha256) and an attacker can't trigger padding errors because the plaintext has no protocol-specific lengths or paddings, it's just a blob with a sha.

Is this still unsafe? What do you think?

**Comment by dchest** [comment]
I see, but in this case why even bother with using separate pieces of stream: the purpose of `nacl-stream-js` is to authenticate *every* packet, so that it can discard modified packets early:

https://www.imperialviolet.org/2014/06/27/streamingencryption.html

In your case, you can just use streaming implementation of XSalsa20. It would be faster, because it won't have to expand nonce for every packet. (Also, if care about speed, native non-JS implementation of Poly1305 is much [like 10 times] faster than SHA-256.)

To summarize, unfortunately your software not only brings nothing new to this world, but it discards ~a decade of knowledge we gained about safe crypto systems (discard bad packets as early as possible, use AEAD construction or at least encrypt-then-MAC, use high-level constructions).

(PS I'm writing this in good faith and hope you won't be offended by my "review".)

**Comment by carlos8f** [comment]
I really appreciate your thoughts, thank you.

> why even bother with using separate pieces of stream

My use-case is processing multi-GB files, larger than would fit into RAM, so I'm interested in doing the minimal authentication per chunk, and keeping memory usage as low as possible.

> In your case, you can just use streaming implementation of XSalsa20.

Referring to `crypto_stream_salsa20`? Would I need to expand the nonce once per stream `crypto_core_hsalsa20(s,n,k,sigma)`, and re-use it for every chunk? Is it unnecessary to increment the nonce after every chunk, then?

> it discards ~a decade of knowledge we gained about safe crypto

I'm still considering things like encrypt-then-mac, but since I have to mac-then-encrypt anyway to ensure the proper output, it would just be a 'pre-check' before decryption happens. Nice to have, but not totally necessary. Since this is not a network protocol I'm not really worried about discarding packets early.

> unfortunately your software not only brings nothing new to this world

My goal is to make a sane alternative to PGP using NaCl. Do you use/know of any CLI tools with public key crypto that can operate on large files? If so I'd love to compare/benchmark.

For salty, I'll be experimenting with all these options (including using nacl-stream-js), and benchmarking. Will get back to you later. Thanks!

**Comment by dchest** [comment]
> Referring to crypto_stream_salsa20?

No, there's no streaming implementation in TweetNaCl-js.

> My goal is to make a sane alternative to PGP using NaCl.

Please read the post I linked earlier on why sane alternative would authenticate each packet.

**Comment by carlos8f** [comment]
@dchest update, I ported salty to use nacl-stream and benchmarked it, and it's about equivalent speed-wise (still faster than gpg!) so I'm keeping it. Thanks for your advice, it helped a lot

**Comment by dchest** [comment]
Awesome!
